### PR TITLE
2021 01 06 tlv invariants

### PR DESCRIPTION
**Pull request by Christewart**
Add invariants to various TLV data structures. Mainly around making sure `Vector`s aren't empty.

@nkohen and i discussed this might be a easy way to get to the bottom of #3869 as when we delete adaptor signatures from the database, we might end up have a `CETSignaturesTLV` that is empty. This gives us the opportunity to "fail-fast" and see the root cause of the issue (not having adaptor sigs).

I haven't tested this out yet on #3869, but this seems like a good idea to add regardless.

I had modify the type returned by `DLCDataManagement.executorAndSetupFromDb()` to return an `Option[DLCExecutorWithSetup]`. The `None` case represents when we have pruned the cet signatures from the database to save on disk space.

I had to bubble this `Option` throughout the codebase, and try to appropriately handle it. I left comments in the diff where i had questions and could use review there.

### CODE REVIEW

**Review by benthecarman — APPROVED** [review]

**Review by nkohen — CHANGES_REQUESTED** [review]
I'm not sure this will catch what we discussed, and also I don't think that it is necessarily good to have exceptions thrown at this level, the invariants should live on the deserialized versions not the serialized ones (for example, we could have places where we don't go through TLV when communicating between db and wallet). For example, an invariant should probably be on our internal DLCAccept and DLCSign types.

**Comment by Christewart** [comment]
> I'm not sure this will catch what we discussed, and also I don't think that it is necessarily good to have exceptions thrown at this level, the invariants should live on the deserialized versions not the serialized ones (for example, we could have places where we don't go through TLV when communicating between db and wallet). For example, an invariant should probably be on our internal DLCAccept and DLCSign types.

That introduces just a few failures on 8186f14 :sweat_smile: 

Screenshot from 2022-01-06 12-39-18

**Comment by Christewart** [comment]
> > I'm not sure this will catch what we discussed, and also I don't think that it is necessarily good to have exceptions thrown at this level, the invariants should live on the deserialized versions not the serialized ones (for example, we could have places where we don't go through TLV when communicating between db and wallet). For example, an invariant should probably be on our internal DLCAccept and DLCSign types.
> 
> That introduces just a few failures on [8186f14](https://github.com/bitcoin-s/bitcoin-s/commit/8186f1409eaac4824883bcb62fd9400e4cf9b2a3) sweat_smile
> 
> Screenshot from 2022-01-06 12-39-18

Commenting out one line gets all the test cases too pass on 8186f14

https://github.com/bitcoin-s/bitcoin-s/blob/4ca586ca462cd0bf029cfe43f698d9e49da4aa6e/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/internal/DLCTransactionProcessing.scala#L64

My understanding is this line deletes the adaptor signatures in the database. @benthecarman hypothesized this was the root cause of #3869 , still haven't been able to prove it yet but it could be related to this change.

I think i'm going to wrap `Vector[DLCCETSignaturesDb]` -> `Option[Vector[DLCCETSignaturesDb]]` so we have a typesafe way of representing when we do / do not have the cet sigs available to us.

https://github.com/bitcoin-s/bitcoin-s/blob/f71d3567ed21df6bb050ee184d6fe31b4369c654/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/internal/DLCDataManagement.scala#L317

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 1]
```diff
@@ -1272,15 +1272,57 @@ abstract class DLCWallet
...
-      _ = logger.info(
-        s"Created DLC execution transaction ${tx.txIdBE.hex} for contract ${contractId.toHex}")
+            //lets try to retrieve it from our transactionDAO
+            val dlcDbOptF = dlcDAO.findByContractId(contractId)
 
+            for {
+              dlcDbOpt <- dlcDbOptF
+              _ = require(
+                dlcDbOpt.isDefined,
+                s"Could not find dlc associated with this contractId=${contractId}")
```
Comment (benthecarman) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala]: nit: call `.hex` on `contractId`

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 2]
```diff
@@ -1272,15 +1272,57 @@ abstract class DLCWallet
...
+              dlcDb = dlcDbOpt.get
+              _ = require(
+                dlcDb.closingTxIdOpt.isDefined,
+                s"If we don't have CET signatures, the closing tx must be defined, contractId=$contractId")
```
Comment (benthecarman) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala]: nit: call `.hex` on `contractId`

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala [hunk 3]
```diff
@@ -1272,15 +1272,57 @@ abstract class DLCWallet
...
+              closingTxId = dlcDb.closingTxIdOpt.get
+              closingTxOpt <- transactionDAO.findByTxId(closingTxId)
+            } yield {
+              require(
+                closingTxOpt.isDefined,
+                s"Could not find closing tx for DLC in db, contactId=$contractId closingTxId=$closingTxId")
```
Comment (benthecarman) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala]: nit: call `.hex` on `contractId` and `closingTxId`

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/internal/DLCDataManagement.scala [hunk 4]
```diff
@@ -307,14 +307,20 @@ private[bitcoins] trait DLCDataManagement { self: DLCWallet =>
...
-    } yield (dlcDb,
-             contractData,
-             dlcOffer,
-             dlcAccept,
-             refundSigs.head,
-             contractInfo,
-             fundingInputs,
-             outcomeSigs)
+    } yield {
+      logger.info(s"outcomeSigs=$outcomeSigs dlcId=${dlcId.hex}")
```
Comment (benthecarman) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/internal/DLCDataManagement.scala]: `outcomeSigs` could be very large here, maybe make this a trace level log or simplify what it is printing

File: dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/internal/DLCTransactionProcessing.scala [hunk 5]
```diff
@@ -31,50 +32,67 @@ private[bitcoins] trait DLCTransactionProcessing extends TransactionProcessing {
...
+        executorWithSetupOptF.flatMap { case executorWithSetupOpt =>
+          executorWithSetupOpt match {
+            case Some(exeutorWithSetup) =>
+              calculateAndSetStateWithSetupDLC(exeutorWithSetup.setup,
+                                               dlcDb,
+                                               txId)
+            case None =>
+              //what do we do if we don't have the cet sigs?
+              //this means we have already deleted the cet sigs
+              //just return the dlcdb given to us?
```
Comment (benthecarman) [inline dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/internal/DLCTransactionProcessing.scala]: yeah this sounds correct

**Review by benthecarman — COMMENTED** [review]
otherwise lgtm

**Pull request merged**